packages/compass-components/src/components/bson-document-editor.tsx

```diff
@@ -0,0 +1,180 @@
+import type { EditorProps } from './editor';
+import { Editor } from './editor';
+import queryParser from 'mongodb-query-parser';
+import { EJSON } from 'bson';
+import { useElementParentHoverState } from '../utils/use-element-parent-hover-state';
+import { Button } from './leafygreen';
+
+type BSONValue = any;
+
+type BSONDocumentEditorProps = Omit<EditorProps, 'variant'> & {
+  variant?: never;
+  onChangeValue?: (
+    value: BSONValue,
+    error: Error | null,
+    variant: 'EJSON' | 'Shell',
+    text: string,
+    event?: any
+  ) => void;
+} & (
+    | {
+        initialValue?: never;
+      }
+    | {
+        text?: never;
+        initialValue?: BSONValue;
+      }
+  );
+
+const actionsGroupContainer = css({
+  position: 'absolute',
+  display: 'flex',
+  gap: spacing[2],
+  width: '100%',
+  top: spacing[3],
+  right: spacing[3],
+  paddingLeft: spacing[3],
+  paddingRight: spacing[3],
+  pointerEvents: 'none',
+  zIndex: 100,
+});
+
+const actionsGroupItem = css({
+  flex: 'none',
+  pointerEvents: 'all',
+});
+
+function BSONDocumentEditor({
+  initialValue,
+  onChangeValue,
+  text,
+  onChangeText,
+  ...otherProps
+}: BSONDocumentEditorProps): React.ReactElement {
+  const containerRef = useRef<HTMLDivElement | null>(null);
+  const isHovered = useElementParentHoverState(containerRef);
+
+  if (!text && initialValue) {
+    text = queryParser.toJSString(initialValue);
+  }
+
+  const [currentVariant, setCurrentVariant] = useState<'EJSON' | 'Shell'>(
+    'Shell'
+  );
+  const [lastAlternateVariantRepr, setLastAlternateVariantRepr] = useState<{
+    text: string;
+    value: BSONValue;
+  } | null>(null);
+  const [currentText, setCurrentText] = useState(text ?? '');
+
+  function onChangeTextWrapper(text: string, event?: any) {
+    setCurrentText(text);
+    onChangeText?.(text, event);
+    if (onChangeValue) {
+      let value;
+      let error: Error | null = null;
+      try {
+        const parsed = queryParser(currentText);
+        if (!parsed || typeof parsed !== 'object') {
+          // XXX(COMPASS-5689): We've hit the condition in
+          // https://github.com/mongodb-js/ejson-shell-parser/blob/c9c0145ababae52536ccd2244ac2ad01a4bbdef3/src/index.ts#L36
+          // in which instead of returning a parsed value or throwing an error,
+          // the query parser just returns an empty string when encountering
+          // input that can be parsed as JS but not as a valid query.
+          // Unfortunately, this also means that all context around what
+          // caused this error is unavailable here.
+          throw new Error('Field contained invalid input');
+        }
+
+        if (currentVariant === 'Shell') {
+          value = parsed;
+        } else {
+          value = EJSON.deserialize(parsed, { relaxed: false });
+        }
+      } catch (err) {
+        error = err as Error;
+      }
+      onChangeValue(value, error, currentVariant, text, event);
+    }
+  }
+
+  function onToggle() {
```

Comment:
An interesting corner case here: if I first break syntax, switch, then fix the syntax and switch again, the switcher goes "out of sync" with content and transforms the value into an unexpected format

![Kapture 2022-08-04 at 14 00 10](https://user-images.githubusercontent.com/5036933/182841998-c034de2b-a3b1-46ff-8a16-8ad3c5cd024b.gif)